### PR TITLE
Use Geant4 environment variables to choose run manager and threads

### PR DESCRIPTION
**Pull request by sethrj**
Three main changes here:
- Deleted the `NUM_THREADS` in favor of `G4FORCE_RUN_MANAGER_TYPE` and `G4FORCENUMBEROFTHREADS`. (Since the default is "Tasking" instead of "MT" I don't use the `Default` enum, so the `FORCE_RUN...` must be used.)
- This let me figure out that I forgot to add the "along-step" initialization for serial mode: so I unified the `RunAction` and `MasterRunAction` by setting a flag whether an instance is responsible for setting up the shared parameters (true for master thread or if serial)
- Because there were confusing asymmetries in Initialize/Finalize I added an `Initialize` to LocalTransporter and made `Initialize` for `SharedParams` only valid on the "master" thread. Other thread-local initialization (i.e. CUDA setup) is done with a separate `InitializeWorker` function.
- With this change, there is no longer any threading assumptions in `accel`, which should give us better compatibility with tasking/TBB use cases.

### CODE REVIEW

**Comment by amandalund** [comment]
I think since the CUDA initialization is only done by the worker threads now the `CoreParams` device ref isn't being created:
```
-------- EEEE ------- G4Exception-START -------- EEEE -------
*** G4Exception : celer0001
      issued by : celeritas/global/CoreParams.hh:129
postcondition failed: device_ref_
*** Fatal Exception *** core dump ***
G4WT6 >  **** Track information is not available at this moment
G4WT6 >  **** Step information is not available at this moment
G4WT6 > 
-------- EEEE -------- G4Exception-END --------- EEEE -------
```

**Comment by sethrj** [comment]
Whoops I meant to call that function from the constructor as well!

**Review by amandalund — APPROVED** [review]
Looking good now @sethrj!

File: src/accel/LocalTransporter.hh [hunk 1]
```diff
@@ -71,5 +75,18 @@ class LocalTransporter
     size_type max_steps_{};
 };
 
+//---------------------------------------------------------------------------//
+/*!
+ * Helper for making initialization more obvious from user code.
+ *
+ * Since a LocalTransporter is shared across threads, this gives it some
+ * symmetry with Finalize.
```
[the remark this answers is not in the document]
Comment (sethrj) [reply src/accel/LocalTransporter.hh:83]: Whoops, looks like a copy paste error on my part.

File: src/accel/SharedParams.cc [hunk 2]
```diff
@@ -153,9 +166,11 @@ void SharedParams::Finalize()
 
 //---------------------------------------------------------------------------//
 /*!
- * Construct from setup options in a thread-safe manner.
+ * Construct from setup options.
+ *
+ * This is not thread-safe and should
```
[the remark this answers is not in the document]
Comment (sethrj) [reply src/accel/SharedParams.cc:171]: 🐿️

File: src/accel/SharedParams.hh [hunk 3]
```diff
@@ -30,8 +30,9 @@ class StepCollector;
  * passed to a thread-local \c LocalTransporter instance. At the beginning of
  * the run, after Geant4 has initialized physics data, the \c Initialize method
  * must be called first on the "master" thread to populate the Celeritas data
- * structures (geometry, physics). \c Initialize must subsequently be invoked
- * on all worker threads.
+ * structures (geometry, physics). \c InitializeWorker must subsequently be
```
Comment (whokion) [inline src/accel/SharedParams.hh:33]: To be specific, physics may be "physics tables" as other properties of physics (for an example, physics processes) are thread-local.

File: src/accel/SharedParams.hh [hunk 4]
```diff
@@ -50,8 +51,14 @@ class SharedParams
...
 
-    // Thread-safe setup of Celeritas using Geant4 data.
-    void Initialize(const SetupOptions& options);
+    // Construct Celeritas using Geant4 data on the master thread.
+    explicit SharedParams(const SetupOptions& options);
+
+    // Initialize shared data on the "master" thread
+    inline void Initialize(const SetupOptions& options);
+
+    // Set up global thread-local data on worker threads
```
Comment (whokion) [inline src/accel/SharedParams.hh:60]: Again, static thread-local data?

**Review by whokion — APPROVED** [review]
I left some minor comments and questions that may be clarified.  Other than that, it is good to go.  Thanks for improving  the code!

File: src/accel/SharedParams.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (amandalund) [reply src/accel/SharedParams.cc:171]: 🐶 🐿️ 🐶 🐿️

File: src/accel/Logger.cc [hunk 5]
```diff
@@ -38,7 +38,7 @@ class MtLogger
 //---------------------------------------------------------------------------//
 MtLogger::MtLogger(int num_threads) : num_threads_(num_threads)
 {
-    CELER_EXPECT(num_threads_ > 0);
+    CELER_EXPECT(num_threads_ >= 0);
```
[the remark this answers is not in the document]
Comment (sethrj) [reply src/accel/Logger.cc:41]: This actually showed up when I accidentally used the "tasking" manager without the explicit call to `SetNumberOfThreads`: it reports the number of threads as zero.

File: src/accel/SharedParams.hh (hunk 3 above)
Comment (sethrj) [reply src/accel/SharedParams.hh:33]: Physics processes aren't thread-local in Celeritas: it's talking about *celeritas* thread-local data structures, not Geant4 ones.

File: app/demo-geant-integration/RunAction.cc [hunk 6]
```diff
@@ -41,25 +43,39 @@ void RunAction::BeginOfRunAction(const G4Run* run)
...
+                = GlobalSetup::Instance()->GetGeometryFile();
+        }
 
-            // Construct thread-local transporter
-            *transport_ = celeritas::LocalTransporter(*options_, *params_);
-            CELER_ENSURE(*transport_);
-        },
-        call_g4exception);
+        // Create the along-step action
+        GlobalSetup::Instance()->SetAlongStep(NoFieldAlongStepFactory{});
```
[the remark this answers is not in the document]
Comment (sethrj) [reply app/demo-geant-integration/RunAction.cc:60]: See the discussion in #586: this particular iteration of the geant app uses no field, but we should enhance the app with a field for testing. As a follow-up PR, either you or I should move/extend the one of the nonuniform field classes in `test/celeritas/field`: either copying the CMS parameterized field into the demo app, or polishing and moving the field map reader.

File: app/demo-geant-integration/RunAction.cc [hunk 7]
```diff
@@ -71,14 +87,17 @@ void RunAction::EndOfRunAction(const G4Run*)
     CELER_LOG_LOCAL(status) << "Finalizing Celeritas";
     celeritas::ExceptionConverter call_g4exception{"celer0005"};
 
-    // Deallocate Celeritas state data (ensures that objects are deleted on the
-    // thread in which they're created, necessary by some geant4
-    // thread-local allocators)
-    CELER_TRY_ELSE(transport_->Finalize(), call_g4exception);
+    if (transport_)
+    {
+        // Deallocate Celeritas state data (ensures that objects are deleted on
```
[the remark this answers is not in the document]
Comment (sethrj) [reply app/demo-geant-integration/RunAction.cc:92]: The `thread_local` comment here is in regards to [the bug Amanda encountered](https://github.com/celeritas-project/celeritas/pull/581#issuecomment-1366698686) and [I diagnosed](https://github.com/celeritas-project/celeritas/pull/581#issuecomment-1369007816):

> `G4TouchableHistory` has a special *thread-local* allocator that's created in `operator new` and referenced in `operator delete`. I think this means that it *must* be deleted in the same thread that it's created? I wonder if what's happening is that its allocation (i.e. wherever we first create the `HitProcessor`) is happening on a different thread than the deallocation (from `~HitProcessor`, whenever the last bits of celeritas destruct).

I'm guessing there are other `thread_local`-storage allocators and `delete` overrides hidden throughout Geant4, so it seems safer to guarantee that all the objects we create are destroyed on the same thread.

**Pull request merged**